File: backend/src/main/java/com/credx/match/service/MatchEngineService.java
package com.credx.match.service;

import com.credx.match.dto.JobDTO;
import com.credx.match.dto.JobMatchDTO;
import com.credx.match.dto.StudentProfileDTO;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class MatchEngineService {

    private final Map<String, String> SKILL_ROADMAPS = Map.of(
            "Angular", "https://angular.dev/tutorials",
            "Spring Boot", "https://spring.academy/courses",
            "TypeScript", "https://www.typescriptlang.org/docs/",
            "RxJS", "https://rxjs.dev/guide/overview",
            "SQL", "https://sqlbolt.com/",
            "Docker", "https://docs.docker.com/get-started/",
            "Figma", "https://help.figma.com/hc/en-us"
    );

    public JobMatchDTO calculateMatch(StudentProfileDTO student, JobDTO job) {
        List<String> required = job.getSkillsRequired() != null ? job.getSkillsRequired() : Collections.emptyList();
        List<String> studentSkills = student.getSkills() != null ? student.getSkills() : Collections.emptyList();

        List<String> matchingSkills = new ArrayList<>();
        List<String> missingSkills = new ArrayList<>();

        for (String req : required) {
            boolean hasSkill = studentSkills.stream().anyMatch(s -> s.equalsIgnoreCase(req));
            if (hasSkill) {
                matchingSkills.add(req);
            } else {
                missingSkills.add(req);
            }
        }

        // 1. Skill Score (50%)
        double skillRatio = required.isEmpty() ? 1.0 : (double) matchingSkills.size() / required.size();
        double skillScore = skillRatio * 50.0;

        // 2. GPA Score (30%)
        double gpaThreshold = job.getGpaThreshold() != null ? job.getGpaThreshold() : 0.0;
        double studentGpa = student.getGpa() != null ? student.getGpa() : 0.0;
        double gpaScore = 30.0;
        if (gpaThreshold > 0 && studentGpa < gpaThreshold) {
            gpaScore = (studentGpa / gpaThreshold) * 30.0;
        }

        // 3. Work Authorization Score (20%)
        boolean jobSponsors = Boolean.TRUE.equals(job.getSponsorshipRequired());
        boolean studentAuthorized = "authorized".equalsIgnoreCase(student.getWorkAuthorization());
        double workAuthScore = 20.0;

        if (!studentAuthorized && !jobSponsors) {
            workAuthScore = 0.0;
        }

        int overallScore = (int) Math.round(skillScore + gpaScore + workAuthScore);
        overallScore = Math.min(Math.max(overallScore, 35), 98); // Clamp between 35 and 98

        String explanation;
        if (overallScore >= 85) {
            explanation = "Exceptional fit! High technical skill alignment (" + matchingSkills.size() + "/" + required.size() + " skills) and meets GPA threshold (" + studentGpa + "/" + gpaThreshold + ").";
        } else if (overallScore >= 65) {
            explanation = "Moderate compatibility. Meets academic and work auth criteria. Acquire " + missingSkills.stream().collect(Collectors.joining(", ")) + " to maximize score.";
        } else {
            explanation = "Low compatibility. Requires additional technical skill alignment in " + missingSkills.stream().collect(Collectors.joining(", ")) + ".";
        }

        List<String> learningResources = missingSkills.stream()
                .map(skill -> skill + " Guide: " + SKILL_ROADMAPS.getOrDefault(skill, "https://developer.mozilla.org/"))
                .collect(Collectors.toList());

        return new JobMatchDTO(job, overallScore, matchingSkills, missingSkills, explanation, learningResources);
    }
}

File: src/app/core/services/match-engine.service.ts
import { Injectable } from '@angular/core';
import { StudentProfile } from '../models/profile.model';
import { Job } from '../models/job.model';

export interface LearningResource {
  skill: string;
  resource: string;
  link: string;
}

export interface MatchBreakdown {
  score: number;
  skillsScore: number;
  skillsMatched: string[];
  skillsMissing: string[];
  skillsPercent: number;
  gpaScore: number;
  gpaPassed: boolean;
  gpaMargin: number;
  workAuthScore: number;
  workAuthCompatible: boolean;
  explanation: string[];
  roadmap: LearningResource[];
}

@Injectable({
  providedIn: 'root'
})
export class MatchEngineService {

  private readonly RESOURCE_MAP: Record<string, { resource: string; link: string }> = {
    'angular': { resource: 'Angular official tutorials & guides', link: 'https://angular.dev' },
    'typescript': { resource: 'TypeScript official handbook & playground', link: 'https://www.typescriptlang.org' },
    'rxjs': { resource: 'RxJS Operators interactive reference guide', link: 'https://rxjs.dev' },
    'scss': { resource: 'Sass Guide & normalizer structures', link: 'https://sass-lang.com' },
    'java': { resource: 'Oracle Java tutorials and documentation', link: 'https://dev.java' },
    'spring boot': { resource: 'Spring Academy - Build Spring Boot REST APIs', link: 'https://spring.academy' },
    'sql': { resource: 'PostgreSQL official tutorial manual', link: 'https://www.postgresql.org' },
    'rest api': { resource: 'RESTful API Design guidelines on MDN', link: 'https://developer.mozilla.org' },
    'docker': { resource: 'Docker Curriculum - Complete container course', link: 'https://docker-curriculum.com' },
    'figma': { resource: 'Figma Design Academy for developers', link: 'https://www.figma.com/resources/learn-design' },
    'git': { resource: 'Pro Git Book (free interactive edition)', link: 'https://git-scm.com/book' },
    'web performance': { resource: 'web.dev performance metrics & indicators', link: 'https://web.dev/performance-scoring' },
    'animations': { resource: 'CSS Tricks Complete Guide to CSS Transitions', link: 'https://css-tricks.com/almanac/properties/t/transition' }
  };

  calculateMatch(profile: StudentProfile | null, job: Job): MatchBreakdown {
    // If no profile, return 0 score and default breakdown
    if (!profile) {
      return {
        score: 0,
        skillsScore: 0,
        skillsMatched: [],
        skillsMissing: job.skillsRequired,
        skillsPercent: 0,
        gpaScore: 0,
        gpaPassed: false,
        gpaMargin: 0,
        workAuthScore: 0,
        workAuthCompatible: false,
        explanation: ['Create a profile to calculate your matching compatibility.'],
        roadmap: this.generateRoadmap(job.skillsRequired)
      };
    }

    const explanation: string[] = [];
    
    // 1. Skill Overlap (Weight: 50 points)
    const requiredSkills = job.skillsRequired || [];
    const studentSkills = profile.skills || [];
    
    let matchedSkills: string[] = [];
    let missingSkills: string[] = [];
    let skillsScore = 0;
    let skillsPercent = 0;

    if (requiredSkills.length > 0) {
      matchedSkills = requiredSkills.filter(skill => 
        studentSkills.some(studentSkill => studentSkill.toLowerCase() === skill.toLowerCase())
      );
      missingSkills = requiredSkills.filter(skill => 
        !studentSkills.some(studentSkill => studentSkill.toLowerCase() === skill.toLowerCase())
      );
      
      skillsPercent = Math.round((matchedSkills.length / requiredSkills.length) * 100);
      skillsScore = Math.round((matchedSkills.length / requiredSkills.length) * 50);
      
      if (skillsPercent === 100) {
        explanation.push(`You match all required skills for this role.`);
      } else if (skillsPercent >= 50) {
        explanation.push(`Strong skills overlap: You match ${matchedSkills.length} of ${requiredSkills.length} skills.`);
      } else if (matchedSkills.length > 0) {
        explanation.push(`Partial skills overlap: You are missing ${missingSkills.join(', ')}.`);
      } else {
        explanation.push(`Skill gap: This role requires skills you haven't listed yet.`);
      }
    } else {
      skillsPercent = 100;
      skillsScore = 50;
      explanation.push(`No specific technical skills are required.`);
    }

    // 2. GPA Threshold (Weight: 30 points)
    let gpaScore = 0;
    let gpaPassed = false;
    const gpaMargin = profile.gpa - job.gpaThreshold;

    if (profile.gpa >= job.gpaThreshold) {
      gpaScore = 30;
      gpaPassed = true;
      explanation.push(`Your GPA of ${profile.gpa.toFixed(2)} meets the minimum threshold of ${job.gpaThreshold.toFixed(2)}.`);
    } else if (gpaMargin >= -0.3) {
      gpaScore = 15;
      gpaPassed = false;
      explanation.push(`Your GPA of ${profile.gpa.toFixed(2)} is close to the threshold of ${job.gpaThreshold.toFixed(2)}.`);
    } else {
      gpaScore = 0;
      gpaPassed = false;
      explanation.push(`Your GPA of ${profile.gpa.toFixed(2)} is below the target requirement of ${job.gpaThreshold.toFixed(2)}.`);
    }

    // 3. Work Auth Compatibility (Weight: 20 points)
    let workAuthScore = 0;
    let workAuthCompatible = false;

    if (profile.workAuthorization === 'authorized') {
      workAuthScore = 20;
      workAuthCompatible = true;
      explanation.push(`Authorized to work: You don't require visa sponsorship.`);
    } else {
      if (job.sponsorshipRequired) {
        workAuthScore = 20;
        workAuthCompatible = true;
        explanation.push(`Sponsorship available: The employer supports candidates requiring visa sponsorship.`);
      } else {
        workAuthScore = 0;
        workAuthCompatible = false;
        explanation.push(`Sponsorship gap: The employer requires existing US work authorization.`);
      }
    }

    const totalScore = skillsScore + gpaScore + workAuthScore;
    const roadmap = this.generateRoadmap(missingSkills);

    return {
      score: totalScore,
      skillsScore,
      skillsMatched: matchedSkills,
      skillsMissing: missingSkills,
      skillsPercent,
      gpaScore,
      gpaPassed,
      gpaMargin,
      workAuthScore,
      workAuthCompatible,
      explanation,
      roadmap
    };
  }

  private generateRoadmap(missingSkills: string[]): LearningResource[] {
    return missingSkills.map(skill => {
      const match = this.RESOURCE_MAP[skill.toLowerCase()];
      if (match) {
        return { skill, ...match };
      }
      return { 
        skill, 
        resource: `Coursera / Udemy courses on ${skill}`, 
        link: `https://www.coursera.org/search?query=${encodeURIComponent(skill)}` 
      };
    });
  }
}
